internal/target/repository.go

```diff
@@ -237,52 +237,69 @@ func (r *Repository) FetchAuthzProtectedEntitiesByScope(ctx context.Context, pro
+	query := fmt.Sprintf(listTargets, limit)
+	args := []any{}
+	if opts.WithStartPageAfterItem != nil {
+		query = fmt.Sprintf(listTargetsPage, limit)
+		args = append(args,
+			sql.Named("last_item_create_time", opts.WithStartPageAfterItem.GetCreateTime()),
+			sql.Named("last_item_id", opts.WithStartPageAfterItem.GetPublicId()),
+		)
 	}
-	whereClause := strings.Join(where, " or ")
+
+	return r.queryTargets(ctx, query, args)
+}
+
+// listTargetsRefresh lists targets in a project based on the data in the WithPermissions option
+// provided to the Repository constructor. If no permissions are available, this function
+// is a no-op.
+// Supported options:
+//   - WithLimit which overrides the limit set in the Repository object
+//   - WithStartPageAfterItem which sets where to start listing from
+func (r *Repository) listTargetsRefresh(ctx context.Context, updatedAfter time.Time, opt ...Option) ([]Target, error) {
+	const op = "target.(Repository).listTargetsRefresh"
+
 	opts := GetOpts(opt...)
 	limit := r.defaultLimit
 	if opts.WithLimit != 0 {
 		limit = opts.WithLimit
 	}
-	// Ordering and pagination are tightly coupled.
-	// We order by update_time ascending so that new
-	// and updated items appear at the end of the pagination.
-	// We need to further order by public_id to distinguish items
-	// with identical update times.
-	withOrder := "update_time asc, public_id asc"
+
+	query := fmt.Sprintf(refreshTargets, limit)
+	args := []any{
+		sql.Named("updated_after_time", timestamp.New(updatedAfter)),
+	}
 	if opts.WithStartPageAfterItem != nil {
-		// Now that the order is defined, we can use a simple where
-		// clause to only include items updated since the specified
-		// start of the page. We use greater than or equal for the update
-		// time as there may be items with identical update_times. We
-		// then use public_id as a tiebreaker.
+		query = fmt.Sprintf(refreshTargetsPage, limit)
 		args = append(args,
-			sql.Named("after_item_update_time", opts.WithStartPageAfterItem.GetUpdateTime()),
-			sql.Named("after_item_id", opts.WithStartPageAfterItem.GetPublicId()),
+			sql.Named("last_item_update_time", opts.WithStartPageAfterItem.GetUpdateTime()),
+			sql.Named("last_item_id", opts.WithStartPageAfterItem.GetPublicId()),
 		)
-		whereClause = "(" + whereClause + ") and (update_time > @after_item_update_time or (update_time = @after_item_update_time and public_id > @after_item_id))"
 	}
 
-	var foundTargets []*targetView
-	err := r.reader.SearchWhere(
-		ctx,
-		&foundTargets,
-		whereClause,
-		args,
-		db.WithLimit(limit),
-		db.WithOrder(withOrder),
-	)
+	return r.queryTargets(ctx, query, args)
+}
+
+func (r *Repository) queryTargets(ctx context.Context, query string, args []any) ([]Target, error) {
+	const op = "target.(Repository).queryTargets"
+
+	rows, err := r.reader.Query(ctx, query, args)
 	if err != nil {
 		return nil, errors.Wrap(ctx, err, op)
 	}
-
+	defer rows.Close()
+	var foundTargets []*targetView
```

Comment:
I'd like to draw attention to how this listing works, because it is currently unique to Targets among all our resources:
1. A view is created that uses an SQL union to append all entries for all target subtypes to a single "table", carefully crafting the union such that table rows are uniformly typed.
2. We create a type (`targetView`) that can read the structure of the view and can be used to scan the result from the query
3. We loop over each entry in the view and, depending on the "type", further unmarshal the value using a registered subtype handler. The final slice is one of the top level resource interface type.
This structure allows the top level ("base") package, in this case `internal/target` to perform the query and scan the results without having specific knowledge of how to scan the subtype rows. My question is this - do we want to apply this pattern to other resource that have a similar sort of structure? Examples include auth methods, credential stores, host catalogs, etc. It would require creating a subtype registry for all of these domain packages and using this registration model. Personally I'd love to _avoid_ this, but I'm not sure how else to solve this problem, since we have to very carefully avoid an import cycle.

Reply:
We can probably look more closely when we get to one of those packages. Personally I think it is a good pattern, and there is no need to avoid it. But then again, I did implement it in the target package :shrug:.